File: leetcode/141_linked_list_cycle/solution.ts
/**
 * Definition for singly-linked list.
 */

class ListNode {
  val: number;
  next: ListNode | null;
  constructor(val?: number, next?: ListNode | null) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
  }
}

function hasCycle(head: ListNode | null): boolean {
  const TOUCHED_VAL: number = 10 ** 6; // 10**-5 <= node.val <= 10**5

  let node: ListNode = head;

  while (true) {
    if (!node) {
      break;
    }

    if (node.val === TOUCHED_VAL) {
      return true;
    }

    node.val = TOUCHED_VAL;
    node = node.next;
  }

  return false;
}

File: leetcode/141_linked_list_cycle/solution.py
# Definition for singly-linked list.
from typing import Optional


class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


# first attempt
class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        reached = set()
        node = head

        while True:
            if not node:
                break

            if node in reached:
                return True

            reached.add(node)
            node = node.next

        return False


# try to reduce memory usage
class Solution2:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        node = head
        TOUCHED_VALUE = 10**6  # -10**5 <= Node.val <= 10**5 , 10**6 is impossible to appear in the node

        while True:
            if not node:
                break

            if node.val == TOUCHED_VALUE:
                return True

            node.val = TOUCHED_VALUE
            node = node.next

        return False


